react-native-git-upgrade/cliEntry.js

```diff
@@ -0,0 +1,239 @@
+/**
+ * Copyright (c) 2015-present, Facebook, Inc.
+ * All rights reserved.
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree. An additional grant
+ * of patent rights can be found in the PATENTS file in the same directory.
+ */
+'use strict';
+
+import fs from 'fs';
+import os from 'os';
+import path from 'path';
+import shell from 'shelljs';
+import Promise from 'promise';
+import yeoman from 'yeoman-environment';
+import TerminalAdapter from 'yeoman-environment/lib/adapter'
+import log from 'npmlog';
+import rimraf from 'rimraf';
+
+import {
+  checkDeclaredVersion,
+  checkMatchingVersions,
+  checkReactPeerDependency,
+  checkGitAvailable,
+  checkNewVersion
+} from './checks';
+
+log.heading = 'git-upgrade';
+
+/**
+ * Promisify the callback-based shelljs function exec
+ * @param command
+ * @param opts
+ * @returns {Promise}
+ */
+function exec(context, command, forceOutput) {
+  return new Promise((resolve, reject) => {
+    shell.exec(command, {silent: (forceOutput ? !forceOutput : !context.cliArgs.verbose)}, (code, stdout, stderr) => {
+      code
+        ? reject(new Error(`Command '${command}' exited with code ${code}:
+stderr: ${stderr}
+stdout: ${stdout}`))
+        : resolve(stdout);
+    });
+  });
+}
+
+function readProjectDescriptors() {
+  try {
+    const rnPakPath = path.resolve(
+      process.cwd(),
+      'node_modules',
+      'react-native',
+      'package.json'
+    );
+
+    const pakPath = path.resolve(
+      process.cwd(),
+      'package.json'
+    );
+
+    const rnPak = JSON.parse(fs.readFileSync(rnPakPath, 'utf8'));
+    const pak = JSON.parse(fs.readFileSync(pakPath, 'utf8'));
+
+    return {rnPak, pak};
+  } catch (err) {
+    throw new Error(
+      'Unable to find project descriptors. Make sure that you have run `npm install` ' +
+      'and that you are inside a react-native project.'
+    )
+  }
+}
+
+
+function setupWorkingDir(context) {
+  return new Promise((resolve, reject) => {
+    rimraf(context.tmpDir, err => {
+      if (err) {
+        reject(err);
+      } else {
+        fs.mkdirSync(context.tmpDir);
+        resolve()
+      }
+    });
+  });
+}
+
+function configureGitEnv(context) {
+  process.env.GIT_DIR = path.resolve(context.tmpDir, '.gitrn');
+  process.env.GIT_WORK_TREE = '.';
+}
+
+function generateTemplates(context) {
+  try {
+    // Try requiring the index.js (entry-point of Yeoman generators)
+    const generatorEntryPoint = path.resolve(context.generatorDir, 'index.js');
+    fs.accessSync(generatorEntryPoint);
+    return runYeomanGenerators(context);
+  } catch(err) {
+    return runCopyAndReplace(context);
+  }
+}
+
+function runCopyAndReplace(context) {
+  const copyProjectTemplateAndReplacePath = path.resolve(context.generatorDir, 'copyProjectTemplateAndReplace');
+  // This module can be required twice in different RN version
```

Comment:
What do you think of:
```javascript
  /*
   * This module is required twice during the process: for both old and new version
   * of React Native.
   * This file could have changed between these 2 versions. When generating the new template, 
   * we don't want to load the old version of the generator from cache
   */
```

Note: It is unused for now because the old template will be generated from Yeoman so the `copyProjectTemplateAndReplace` will be call only once for the new template. But from the 0.38.0 to the future 0.39.0, 
the `copyProjectTemplateAndReplace` will be required and called to generate the 0.38.0 template, and then it will be required and called to generate the 0.39.0. 
The problem is: the `require` acts as a singleton, any further `require(something...)` calls will return the cached result of the first one. If we don't kill the cache between the 2 `require` calls, we'll generate the 0.39.0 template with the 0.38.0 generator.